### PR TITLE
Allow multiple message arguments for static assert

### PR DESCRIPTION
**Pull request by ntrel**
Reboot of #11757 by @thewilsonator.

### CODE REVIEW

File: compiler/src/dmd/astbase.d [hunk 1]
```diff
@@ -441,14 +441,15 @@ struct ASTBase
...
+        Expressions* msg;
 
         extern (D) this(const ref Loc loc, Expression exp, Expression msg)
         {
             super(Id.empty);
             this.loc = loc;
             this.exp = exp;
-            this.msg = msg;
+            this.msg = new Expressions(1);
+            this.msg.push(msg);
```
Comment (tgehr) [inline compiler/src/dmd/astbase.d:452]: Why only one constructor in ASTBase?

**Comment by tgehr** [comment]
I guess one thing that is a bit weird is that this does not expand sequences:
```d
alias Seq(T...)=T;
static assert(false, Seq!("123", "456"));
```
Idk if it's desirable for `pragma(msg, ...)`, but for `static assert` it may be confusing. In either case, probably good to add an explicit test case for this to document the intention.

**Comment by ntrel** [comment]
@tgehr I just tried calling `expandTuples(sa.msg);` but it doesn't change the array. It has no effect for:
```d
static assert(false, Seq!("123", 456));
```

File: compiler/src/dmd/astbase.d (hunk 1 above)
Comment (ntrel) [reply compiler/src/dmd/astbase.d]: Added `Expressions* msg` ctor.

**Comment by thewilsonator** [comment]
> I guess one thing that is a bit weird is that this does not expand sequences:

Yeah, Ideally this should behave the same way as `pragma(msg, bunch, "of",argu+ments)`

File: compiler/src/dmd/semantic2.d [hunk 2]
```diff
@@ -112,19 +112,39 @@ private extern(C++) final class Semantic2Visitor : Visitor
...
-                const slice = se.toUTF8(sc).peekString();
-                error(sa.loc, "static assert:  \"%.*s\"", cast(int)slice.length, slice.ptr);
+                Expression e = (*sa.msg)[i];
+                sc = sc.startCTFE();
+                e = e.expressionSemantic(sc);
+                e = resolveProperties(sc, e);
+                sc = sc.endCTFE();
+                if (e.type && e.type.ty == Tvoid)
+                {
+                    error(sa.loc, "cannot pass argument `%s` to `static assert` because it is `void`", e.toChars());
```
Comment (thewilsonator) [inline compiler/src/dmd/semantic2.d:125]: Why? `pragma(msg)` can print types and this artificially disallows void but not other types.

Comment (ntrel) [reply compiler/src/dmd/semantic2.d]: Removed. It came from here:
https://github.com/dlang/dmd/pull/11757/files#diff-0ebb96970f91dc91b6b08bcecc0fbd18972450e09bae64df7e87621e09342c79R120

**Comment by ntrel** [comment]
@thewilsonator 
> > I guess one thing that is a bit weird is that this does not expand sequences:
> 
> Yeah, Ideally this should behave the same way as `pragma(msg, bunch, "of",argu+ments)`

This does behave the same:
```d
alias Seq(T...) = T;
pragma(msg, Seq!(int, 0)); // tuple((int), 0)
```

**Comment by ntrel** [comment]
> @tgehr I just tried calling `expandTuples(sa.msg);` but it doesn't change the array. It has no effect for:
> 
> ```d
> static assert(false, Seq!("123", 456));
> ```

I found out that's because the Seq instantiation gets parsed as a ScopeExp, not a TupleExp (or TypeExp) which expandTuples expects.

File: compiler/src/dmd/hdrgen.d [hunk 3]
```diff
@@ -3460,7 +3463,7 @@ private void dumpTemplateInstance(TemplateInstance ti, OutBuffer* buf, HdrGenSta
     }
     else if (ti.members)
     {
-        foreach(m;*ti.members)
+        foreach (m; *ti.members)
```
Comment (RazvanN7) [inline compiler/src/dmd/hdrgen.d:3466]: Please leave do not mix implementations with refactorings.

File: compiler/src/dmd/astbase.d [hunk 4]
```diff
@@ -442,12 +442,19 @@ struct ASTBase
     extern (C++) final class StaticAssert : Dsymbol
     {
         Expression exp;
-        Expression msg;
+        Expressions* msg;
```
Comment (RazvanN7) [inline compiler/src/dmd/astbase.d:445]: Maybe call it `msgs`?

Comment (RazvanN7) [reply compiler/src/dmd/astbase.d:445]: If this requires many changes it could be left for a subsequent PR.

**Review by RazvanN7 — APPROVED** [review]

File: compiler/src/dmd/astbase.d (hunk 4 above)
Comment (ntrel) [reply compiler/src/dmd/astbase.d:445]: @RazvanN7 now done and removed style changes. Also there is a spec PR.

**Comment by RazvanN7** [comment]
Looks good to me. cc @WalterBright

**Comment by RazvanN7** [comment]
@ntrel I assume @WalterBright is busy with other things but I don't think he would oppose this addition. I'm going to go ahead and merge this.

**Comment by thewilsonator** [comment]
Thanks for getting this in. Should we add a changelog entry for this?

**Comment by ntrel** [comment]
@thewilsonator Thanks for making it originally ;-) The changelog was included:
https://dlang.org/changelog/pending.html#dmd.static-assert

**Pull request merged**